pkg/cleaner/cleaner.go

```diff
@@ -0,0 +1,373 @@
+)
+
+var isPVCSubPathCleanerRunning = false
+
+type PVCSubPathCleaner struct {
+	pipelineRunApi v1beta1.PipelineRunInterface
+	subPathStorage *storage.PVCSubPathsStorage
+	clientset      *kubernetes.Clientset
+	conf           *k8s.PVCCleanerConfig
+	namespace      string
+}
+
+func NewPVCSubPathCleaner(pipelineRunApi v1beta1.PipelineRunInterface, subPathStorage *storage.PVCSubPathsStorage, clientset *kubernetes.Clientset, conf *k8s.PVCCleanerConfig, namespace string) *PVCSubPathCleaner {
+	return &PVCSubPathCleaner{
+		pipelineRunApi: pipelineRunApi,
+		subPathStorage: subPathStorage,
+		clientset:      clientset,
+		conf:           conf,
+		namespace:      namespace,
+	}
+}
+
+func (cleaner *PVCSubPathCleaner) ScheduleCleanUpSubPathFoldersContent() {
+	for {
+		time.Sleep(DEF_PERIOD)
+
+		log.Println("------------ Schedule cleanup new subpath folders content")
+		if err := cleaner.cleanUpSubPathFoldersContent(); err != nil {
+			log.Print(err)
+		}
+	}
+}
+
+func (cleaner *PVCSubPathCleaner) WatchNewPipelineRuns(storage *storage.PVCSubPathsStorage) {
+	resourceVersion, err := cleaner.conf.GetWatchResourceVersion()
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("************ Resource version for watch add operation is %s", resourceVersion)
+
+	// Watcher will be closed after some timeout, so we need to re-create watcher https://github.com/kubernetes/client-go/issues/623.
+	// Let's use "NewRetryWatcher" helper for this purpose.
+	retryWatcher, err := watchTool.NewRetryWatcher(resourceVersion, &cache.ListWatch{
+		WatchFunc: func() func(options metav1.ListOptions) (watchapi.Interface, error) {
+			return func(options metav1.ListOptions) (watchapi.Interface, error) {
+				return cleaner.pipelineRunApi.Watch(context.TODO(), metav1.ListOptions{})
+			}
+		}(),
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for {
+		event, ok := <-retryWatcher.ResultChan()
+		if !ok {
+			return
+		}
+
+		if event.Type != watchapi.Added {
+			continue
+		}
+		pipelineRun, ok := event.Object.(*pipelinev1.PipelineRun)
+		if !ok {
+			continue
+		}
+		if err = cleaner.conf.UpdateWatchResourceVersion(pipelineRun.ObjectMeta.ResourceVersion); err != nil {
+			log.Println(err)
+			continue
+		}
+
+		log.Printf("************ Add new pipelineRun with name %s", pipelineRun.ObjectMeta.Name)
+
+		var subPath string
+		for _, workspace := range pipelineRun.Spec.Workspaces {
+			if workspace.Name == "workspace" {
+				subPath = workspace.SubPath
+				break
+			}
+		}
+
+		if subPath != "" {
+			pvcSubPath := &model.PVCSubPath{PipelineRun: pipelineRun.ObjectMeta.Name, PVCSubPath: subPath}
+			storage.AddPVCSubPath(pvcSubPath)
+		}
+	}
+}
+
+func (cleaner *PVCSubPathCleaner) WatchAndCleanUpSubPathFolders() {
+	// Watcher will be closed after some timeout, so we need to re-create watcher https://github.com/kubernetes/client-go/issues/623.
+	// Let's use "NewRetryWatcher" helper for this purpose.
+	retryWatcher, err := watchTool.NewRetryWatcher("1", &cache.ListWatch{
```

Comment:
Please add a comment explaining why 1 is suitable here whereas we use stored value for add above

Reply:
Done.